File: app/routers/problems/magicsos.py
from math import log
from statistics import pstdev

from .utils import register_problem, parse_CSV


async def score(square: list[int]):
    # Ensure there are 9 entries
    if len(square) != 9:
        return None
    # Ensure there are distinct elements.
    if len(set(square)) != len(square):
        return None
    # Check if they are all squares.
    for entry in square:
        if entry <= 0 or entry >= 10 ** 20:
            return None
    std = pstdev(square)
    square = [x ** 2 for x in square]

    sums = []
    # Compute the score
    # Compute rows:
    sums.append(square[0] + square[1] + square[2])
    sums.append(square[3] + square[4] + square[5])
    sums.append(square[6] + square[7] + square[8])
    # Compute columns:
    sums.append(square[0] + square[3] + square[6])
    sums.append(square[1] + square[4] + square[7])
    sums.append(square[2] + square[5] + square[8])
    # Compute diagonals:
    sums.append(square[0] + square[4] + square[8])
    sums.append(square[6] + square[4] + square[2])

    M = max(sums)
    m = min(sums)
    result = (M - m) / (log(M) * std)
    return int(result * 10 ** 6)
    

register_problem("magic-square-of-squares", score, "Magic Square of Squares",
                 "magic-square-of-squares.j2", "lowest", "magicsos", 
                 parse_CSV, "magicsos.svg")

File: static/scores/magicsos.ts
import Decimal from "decimal.js";
import * as math from 'mathjs';

export async function score(submission: Decimal[]) {
  try {
    const entries = submission.map((num) => num.toNumber());
    const std = Math.sqrt(math.variance(entries, 'uncorrected') as number);
    console.log(std);
    const squares = entries.map((num) => num * num);
    const sums: number[] = [];
    const dupCheck = new Set<number>(entries);
    if (dupCheck.size != 9) {
      return "Not all numbers are distinct!";
    }
    for (const num of entries) {
      if (num <= 0) {
        return "All number must be positive!";
      }
    }
    // Rows
    sums.push(squares[0] + squares[1] + squares[2]);
    sums.push(squares[3] + squares[4] + squares[5]);
    sums.push(squares[6] + squares[7] + squares[8]);
    // Columns
    sums.push(squares[0] + squares[3] + squares[6]);
    sums.push(squares[1] + squares[4] + squares[7]);
    sums.push(squares[2] + squares[5] + squares[8]);
    // Diagonals
    sums.push(squares[0] + squares[4] + squares[8]);
    sums.push(squares[6] + squares[4] + squares[2]);
 
    const M: number = Math.max(...sums);
    const m: number = Math.min(...sums);
    const result = (M - m) / (math.log(M) * std);
    return math.floor(result * (10 ** 6));
  } catch (e) {
    console.error(e);
    return "Could not score CSV file!";
  }
}
